setup_unit_tests.py

```diff
@@ -0,0 +1,488 @@
+            typ = match.group(1)
+            var_types[name] = typ
+
+    return var_types
+
+
+def wdl_type_to_string(wdl_type: WDL.Type.Base) -> Union[Dict[str, Any], str]:
+    if isinstance(wdl_type, WDL.Type.StructInstance):
+        # this is the only construct that we need to decipher to string form
+        dictified = {}
+        for key, subtype in wdl_type.members.items():
+            dictified[key] = wdl_type_to_string(subtype)
+        return dictified
+    return str(wdl_type)
+
+
+def extract_output_types(wdl_file, fail):
+    """
+    Use miniwdl's parser to extract the output types from a wdl file
+    """
+    if fail:
+        # since we expect this workflow to fail, there should be no outputs
+        return {}
+
+    try:
+        document: WDL.Tree.Document = WDL.load(uri=str(wdl_file))
+    except WDL.Error.InvalidType as e:
+        if "Unknown type Object" in str(e):
+            # one of the tests wants to test if Objects are supported, but miniwdl doesn't support objects
+            # defer to the other more simpler parser
+            return extract_output_types_defer(wdl_file)
+        else:
+            raise RuntimeError("Likely unsupported type. Failed to parse WDL file %s!", wdl_file) from e
+    except Exception as e:
+        raise RuntimeError(f"Failed to parse WDL file {str(wdl_file)}!") from e
+
+    target: Union[WDL.Tree.Workflow, WDL.Tree.Task]
+
+    if document.workflow:
+        target = document.workflow
+    elif len(document.tasks) == 1:
+        target = document.tasks[0]
+    else:
+        raise Exception(
+            "Multiple tasks founds with no workflow!")  # this in theory shouldn't be hit, but the source repo has a suspicious --tasks argument passed to miniwdl sometimes
+
+    var_types = dict()
+    if target.outputs is not None:
+        for output in target.outputs:
+            var_types[output.name] = wdl_type_to_string(output.type)
+
+    return var_types
+
+
+def convert_typed_output_values(output_values: Union[None, str, Dict[str, Any], List[Any]], output_type: WDL.Type.Base,
+                                data_dir: Optional[Path], metadata: Union[List[Any], Dict[str, Any]]):
+    """
+    Get the expected output values with respect to the output types.
+    This is done to get the md5sum of files when converting unit tests.
+
+    Ex: If given hello.txt and type File, try to get the regex of the file
+    If given "hello" and type String, return string
+    """
+    if output_values is None and output_type.optional:
+        return output_values
+    if isinstance(output_type, WDL.Type.File):
+        # this will only populate md5sum if the files exist recursively in the data directory
+        for path in glob.glob(str(data_dir / "**"), recursive=True):
+            if path.endswith(output_values) and os.path.exists(path):
+                with open(path, "rb") as f:
+                    md5sum = hashlib.md5(f.read()).hexdigest()
+                return {'md5sum': md5sum}
```

Comment:
This suggests that we're throwing away information about where exactly the file was supposed to be and then hoping to recover it now. What if two files in the directory hierarchy have the same name but different contents?

Reply:
I intended for this to be closer to a fallback if the key value pair didn't exist. I'll likely get rid of it

Reply:
I didn't get rid of it but moved it to the last option if the patch data does not exist. I don't expect this to be hit too often so I'll keep it for now. I added a comment about this unpredictability